docs/vsock.md

```diff
@@ -0,0 +1,115 @@
+# VSOCK support
+
+VSOCK provides a way for guest and host to communicate through a socket. VSOCK sockets support both stream and datagram types.
+
+The `virtio-vsock` is based on the [Firecracker](https://github.com/firecracker-microvm/firecracker/blob/main/docs/vsock.md) implementation, where additional details can be found.
+
+# Basic Terms
+
+## Predefined Values
+
+| CID | Description |
+|-----|-------------|
+| -1  | Random CID | 
+|  0  | Hypervisor | 
+|  1  | Loopback | 
+|  2  | Host | 
+
+## Socat specific Notes
+
+See the `DATA VALUES` section of `man socat` for the closer info on the data formats. For a quick reference:
+
+| Specifier | Datatype |
+|-----------|----------|
+|     B     | uint8_t  |
+|     S     | uint16_t |
+|     I     | uint32_t |
+
+
+Same man page tells also about the `SOCKET-LISTEN` specificator, where by the last `local-address` component is to contain `the data representation of a sockaddr structure without sa_family`. Consult `linux/vm_sockets.h` on the given system for the details of `struct sockaddr_vm` layout.
+
+Socat `>=1.7.4` supports `VSOCK-CONNECT` and `VSOCK-LISTEN` address types directly, thus simplifying the usage and eliminating the need to pass a manually crafted `sockaddr` struct data.
+
+
+## Kernel Requirements
+
+Host kernel:  CONFIG_VHOST_VSOCK=m
+
+Guest kernel: CONFIG_VIRTIO_VSOCKETS=m 
+
+## Nested VM support
+
+Requirement is Linux `>= 5.5` for the L1 VM.
+
+## Loopback support
+
+Requirement is Linux `>= 5.6`.
+
+# Establishing Vsock Connection
+
+Vsock device becomes available with `--vsock` option passed by the VM start. As per the previous struct listing, both CID and port are 32-bit values.
+
+Cloud Hypervisor can be invoked for instance as below:
+
+```bash
+IMG=jammy-server-cloudimg-amd64.raw
+FV=CLOUDHV_EFI.fd
+CH=cloud-hypervisor
+
+VSOCK_CID=3
+VSOCK_PORT=1234
+VSOCK_SOCK=/tmp/ch.vsock
+
+$CH \
+	--cpus boot=1 \
+	--memory size=4G \
+	--kernel $FV \
+	--disk path=$IMG \
+	--serial tty --console off --net tap= \
+	--vsock cid=$VSOCK_CID,socket=$VSOCK_SOCK
+```
+
+The examples use `socat` to illustrate the VSOCK functionality. However, there are other tools supporting vsock, like [ncat](https://stefano-garzarella.github.io/posts/2019-11-08-kvmforum-2019-vsock/).
+
+## Connecting from Host to Guest
+
+The host starts to listen on the defined port:
+
+`$ socat - SOCKET-LISTEN:40:0:S0I1234I3I0`
+
+With newer `socat` versions, this command can be made simpler:
+
+`$ socat - VSOCK-LISTEN:1234`
+
+Once the host is listenint, the guest can send data:
+
+`echo -e "CONNECT $VSOCK_PORT\\nHello from host!" | socat - UNIX-CONNECT:$VSOCK_SOCK`
+ 
+Note the string `CONNECT <port>` prepended to the actual data. It is possible for the guest to start listening on different ports, thus the specific command is needed to instruct VSOCK to which listener the host wants to connect. It needs to be sent once per a connection. Once the connection established, data transfers can take place directly.
```

Comment:
I think it's quite complex to follow when we know socat 1.7.4 supports VSOCK directly. This would simplify greatly the document.